src/query/storages/common/table-meta/src/meta/v3/snapshot.rs

```diff
@@ -0,0 +1,228 @@
+    pub schema: TableSchema,
+
+    /// Summary Statistics
+    pub summary: Statistics,
+
+    /// Pointers to SegmentInfos (may be of different format)
+    ///
+    /// We rely on background merge tasks to keep merging segments, so that
+    /// this the size of this vector could be kept reasonable
+    pub segments: Vec<Location>,
+
+    // The metadata of the cluster keys.
+    pub cluster_key_meta: Option<ClusterKey>,
+    pub table_statistics_location: Option<String>,
+}
+
+impl TableSnapshot {
+    pub fn new(
+        snapshot_id: SnapshotId,
+        prev_timestamp: &Option<DateTime<Utc>>,
+        prev_snapshot_id: Option<(SnapshotId, FormatVersion)>,
+        schema: TableSchema,
+        summary: Statistics,
+        segments: Vec<Location>,
+        cluster_key_meta: Option<ClusterKey>,
+        table_statistics_location: Option<String>,
+    ) -> Self {
+        let now = Utc::now();
+        // make snapshot timestamp monotonically increased
+        let adjusted_timestamp = util::monotonically_increased_timestamp(now, prev_timestamp);
+
+        // trim timestamp to micro seconds
+        let trimmed_timestamp = util::trim_timestamp_to_micro_second(adjusted_timestamp);
+        let timestamp = Some(trimmed_timestamp);
+
+        Self {
+            format_version: TableSnapshot::VERSION,
+            snapshot_id,
+            timestamp,
+            prev_snapshot_id,
+            schema,
+            summary,
+            segments,
+            cluster_key_meta,
+            table_statistics_location,
+        }
+    }
+
+    pub fn from_previous(previous: &TableSnapshot) -> Self {
+        let id = Uuid::new_v4();
+        let clone = previous.clone();
+        Self::new(
+            id,
+            &clone.timestamp,
+            Some((clone.snapshot_id, clone.format_version)),
+            clone.schema,
+            clone.summary,
+            clone.segments,
+            clone.cluster_key_meta,
+            clone.table_statistics_location,
+        )
+    }
+
+    pub fn format_version(&self) -> u64 {
+        self.format_version
+    }
+
+    pub fn to_bytes(&self) -> Result<Vec<u8>> {
+        let mut buf = Vec::new();
+
+        let encoding = Encoding::default();
+        let compression = Compression::default();
+
+        let data = encode(&encoding, &self)?;
+        let data_compress = compress(&compression, data)?;
+
+        buf.extend_from_slice(&self.format_version.to_le_bytes());
+        buf.push(encoding as u8);
+        buf.push(compression as u8);
+        buf.extend_from_slice(&data_compress.len().to_le_bytes());
+
+        buf.extend(data_compress);
+
+        Ok(buf)
+    }
+
+    #[inline]
+    pub fn encoding() -> Encoding {
+        Encoding::default()
+    }
+}
+
+impl From<v2::TableSnapshot> for TableSnapshot {
+    fn from(s: v2::TableSnapshot) -> Self {
+        Self {
+            format_version: TableSnapshot::VERSION,
```

Comment:
> `FLASHBACK` is not working as expected.

@dantengsky @jun0315 
I'm sorry the previous pr didn't fix the bug, but I've reproduced and tracked the bug and found the problem here. 

Here is an example, snapshot `04015a04838b436689eba248759c6526 `'s version is 2, its format_version should be 2, not 3.
```
mysql> select * from fuse_snapshot('default', 'hits') limit 10;
+----------------------------------+----------------------------------------------------------+----------------+----------------------------------+---------------+-------------+-----------+--------------------+---
---------------+------------+----------------------------+
| snapshot_id                      | snapshot_location                                        | format_version | previous_snapshot_id             | segment_count | block_count | row_count | bytes_uncompressed | by
tes_compressed | index_size | timestamp                  |
+----------------------------------+----------------------------------------------------------+----------------+----------------------------------+---------------+-------------+-----------+--------------------+---
---------------+------------+----------------------------+
| 8ee454392ea44e77b6013ae47d785d47 | 1/367953/_ss/8ee454392ea44e77b6013ae47d785d47_v3.bincode |              3 | 1eb4da851dd740998c8cf7ddae5729ca |             1 |         768 | 128000000 |       107571195904 |
   10375147776 |  763978368 | 2023-04-17 13:20:06.749049 |
| 1eb4da851dd740998c8cf7ddae5729ca | 1/367953/_ss/1eb4da851dd740998c8cf7ddae5729ca_v3.bincode |              3 | 04015a04838b436689eba248759c6526 |            17 |         768 | 128000000 |       107571195904 |
   10375147776 |  763978368 | 2023-04-17 13:19:46.233535 |
| 04015a04838b436689eba248759c6526 | 1/367953/_ss/04015a04838b436689eba248759c6526_v2.json    |              3 | 89f876abd62644c2a9b34b21e3ada816 |             1 |         384 |  64000000 |        53785597952 |
    5187573888 |  381989184 | 2023-04-17 11:10:42.748486 |
    ... ...
+----------------------------------+----------------------------------------------------------+----------------+----------------------------------+---------------+-------------+-----------+--------------------+---
---------------+------------+----------------------------+
10 rows in set (0.17 sec)
Read 10 rows, 2.00 KiB in 0.167 sec., 59.75 rows/sec., 11.97 KiB/sec.
```

Reply:
@zhyass  thanks.  
 
> snapshot 04015a04838b436689eba248759c6526 's version is 2, its format_version should be 2, not 3.

yeah, that is definitely a bug.
 
 but here we are converting a v2::TableSnapshot into v3::TableSansphot, the format version that returned by `fn from(s: v2::TableSnapshot) -> Self ` should be `v3::TableSnapshot::VERSION`, right?

Reply:
> but here we are converting a v2::TableSnapshot into v3::TableSansphot, the format version that returned by `fn from(s: v2::TableSnapshot) -> Self ` should be `v3::TableSnapshot::VERSION`, right?

The version need keep when covert V2 to V3, because the location needs to be generated based on snapshot_id and version. Maybe there is a confusion between snapshot version and format version?

Reply:
So why the format version must be v3 when covert v2 to v3?